README.rst
============================================
GFW Creator: Create FW Perturbation Forcings
============================================


.. image:: https://img.shields.io/pypi/v/gfw-creator.svg
        :target: https://pypi.python.org/pypi/gfw_creator

.. image:: https://github.com/pgierz/gfw_creator/workflows/Python%20package/badge.svg
        :target: https://github.com/pgierz/gfw_creator/actions

.. image:: https://readthedocs.org/projects/gfw-creator/badge/?version=latest
        :target: https://gfw-creator.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status


GFW Creator allows you to generate files for the ``gfw_atmo`` switch in ``ECHAM6/JSBACH`` of ``AWI-ESM``


* Free software: GNU General Public License v3
* Documentation: https://gfw-creator.readthedocs.io.


Features
--------
* Command line utility ``gfw_creator`` to generate suitable forcing files for
  either homogenous freshwater flux or from a reconstruction.
* Plugin for the ``esm-tools`` runtime infrastructure framework.
* Library usage to generate hosing files from other Python scripts.


CHANGELOG.md
## v0.2.1 (2020-12-22)

### Fix

- fixes version numbers, adds commitizen config to automatically increment versions

## v0.2.0 (2020-12-22)

### Fix

- **setup**: adds real cdo dependency
- bumpversion fixup
- fixup version numbers
- fixes unit conversion (it was the wrong way around)
- **gfw_creator**: applies unit conversions to file, assumes user gives Sv

### Feat

- plugin works
- works except for unit conversion


tests/test_gfw_creator.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `gfw_creator` package."""

import pytest

from click.testing import CliRunner

from gfw_creator import gfw_creator
from gfw_creator import cli


@pytest.fixture
def response():
    """Sample pytest fixture.

    See more at: http://doc.pytest.org/en/latest/fixture.html
    """
    # import requests
    # return requests.get('https://github.com/audreyr/cookiecutter-pypackage')


def test_content(response):
    """Sample pytest test function with the pytest fixture as an argument."""
    # from bs4 import BeautifulSoup
    # assert 'GitHub' in BeautifulSoup(response.content).title.string


def test_command_line_interface():
    """Test the CLI."""
    runner = CliRunner()
    result = runner.invoke(cli.main)
    # Running without arguments should give you an error:
    assert result.exit_code == 2
    assert "Usage: main [OPTIONS]" in result.output
    help_result = runner.invoke(cli.main, ["--help"])
    assert help_result.exit_code == 0
    assert "--help  Show this message and exit." in help_result.output


gfw_creator/cli.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
This is the main command line interface for the ``gfw_creator``. You can view
this help by typing either of the following::

    $ man gfw_creator

or::

    $ pydoc gfw_creator.cli

You can create a uniform hosing field over a rectangular area with the
following command::

    $ gfw_creator -- lat_0 lat_1 lon_0 lon_1 hosing_strength

Hosing strength is assumed to be in Sv. lat and lon can be negative, and a -180
to 180 degree domain is assumed for longitude.

The generated output file, ``out.nc`` is created in the current working
directory, and can be used with the AWI-ESM gfw_atmo switch. Below is the result of::

    $ gfw_creator -- 45 60 -30 -10 0.15
    $ ncview out.nc

.. image:: _static/ncview_screenshot.png
  :width: 400
  :alt: GFW Creator output for 45 to 60N, -30 to -10W with 0.15 Sv.
"""
import gfw_creator

import sys
import click


@click.command()
@click.argument("lat_0", type=float)
@click.argument("lat_1", type=float)
@click.argument("lon_0", type=float)
@click.argument("lon_1", type=float)
@click.argument("hosing", type=float)
def main(lat_0, lat_1, lon_0, lon_1, hosing):
    ds = gfw_creator.create_homogeneous_hosing(lat_0, lat_1, lon_0, lon_1, hosing)
    ds.to_netcdf("out.nc")
    return 0


if __name__ == "__main__":
    sys.exit(main())  # pragma: no cover


travis_pypi_setup.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Update encrypted deploy password in Travis config file
"""


from __future__ import print_function
import base64
import json
import os
from getpass import getpass
import yaml
from cryptography.hazmat.primitives.serialization import load_pem_public_key
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric.padding import PKCS1v15


try:
    from urllib import urlopen
except:
    from urllib.request import urlopen


GITHUB_REPO = 'pgierz/gfw_creator'
TRAVIS_CONFIG_FILE = os.path.join(
    os.path.dirname(os.path.abspath(__file__)), '.travis.yml')


def load_key(pubkey):
    """Load public RSA key, with work-around for keys using
    incorrect header/footer format.

    Read more about RSA encryption with cryptography:
    https://cryptography.io/latest/hazmat/primitives/asymmetric/rsa/
    """
    try:
        return load_pem_public_key(pubkey.encode(), default_backend())
    except ValueError:
        # workaround for https://github.com/travis-ci/travis-api/issues/196
        pubkey = pubkey.replace('BEGIN RSA', 'BEGIN').replace('END RSA', 'END')
        return load_pem_public_key(pubkey.encode(), default_backend())


def encrypt(pubkey, password):
    """Encrypt password using given RSA public key and encode it with base64.

    The encrypted password can only be decrypted by someone with the
    private key (in this case, only Travis).
    """
    key = load_key(pubkey)
    encrypted_password = key.encrypt(password, PKCS1v15())
    return base64.b64encode(encrypted_password)


def fetch_public_key(repo):
    """Download RSA public key Travis will use for this repo.

    Travis API docs: http://docs.travis-ci.com/api/#repository-keys
    """
    keyurl = 'https://api.travis-ci.org/repos/{0}/key'.format(repo)
    data = json.loads(urlopen(keyurl).read().decode())
    if 'key' not in data:
        errmsg = "Could not find public key for repo: {}.\n".format(repo)
        errmsg += "Have you already added your GitHub repo to Travis?"
        raise ValueError(errmsg)
    return data['key']


def prepend_line(filepath, line):
    """Rewrite a file adding a line to its beginning.
    """
    with open(filepath) as f:
        lines = f.readlines()

    lines.insert(0, line)

    with open(filepath, 'w') as f:
        f.writelines(lines)


def load_yaml_config(filepath):
    with open(filepath) as f:
        return yaml.load(f)


def save_yaml_config(filepath, config):
    with open(filepath, 'w') as f:
        yaml.dump(config, f, default_flow_style=False)


def update_travis_deploy_password(encrypted_password):
    """Update the deploy section of the .travis.yml file
    to use the given encrypted password.
    """
    config = load_yaml_config(TRAVIS_CONFIG_FILE)

    config['deploy']['password'] = dict(secure=encrypted_password)

    save_yaml_config(TRAVIS_CONFIG_FILE, config)

    line = ('# This file was autogenerated and will overwrite'
            ' each time you run travis_pypi_setup.py\n')
    prepend_line(TRAVIS_CONFIG_FILE, line)


def main(args):
    public_key = fetch_public_key(args.repo)
    password = args.password or getpass('PyPI password: ')
    update_travis_deploy_password(encrypt(public_key, password.encode()))
    print("Wrote encrypted password to .travis.yml -- you're ready to deploy")


if '__main__' == __name__:
    import argparse
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--repo', default=GITHUB_REPO,
                        help='GitHub repo (default: %s)' % GITHUB_REPO)
    parser.add_argument('--password',
                        help='PyPI password (will prompt if not provided)')

    args = parser.parse_args()
    main(args)


gfw_creator/esm_tools_plugin.py
from . import gfw_creator

import sys

from loguru import logger


def create_hosing_files(config):
    if config["general"]["setup_name"] == "awiesm":
        if config["fesom"].get("do_hosing"):
            logger.info("Creating hosing files...")
            hosing_type = config["fesom"].get("hosing_type")
            if hosing_type == "homogeneous":
                lat_0 = config["fesom"]["hosing_lat0"]
                lat_1 = config["fesom"]["hosing_lat1"]
                lon_0 = config["fesom"]["hosing_lon0"]
                lon_1 = config["fesom"]["hosing_lon1"]
                hosing_strength = config["fesom"]["hosing_strength"]
                gfw_atmo_file = gfw_creator.create_homogeneous_hosing(
                    lat_0, lat_1, lon_0, lon_1, hosing_strength
                )
                gfw_atmo_file.to_netcdf(
                    config["fesom"]["thisrun_forcing_dir"] + "/gfw_atmo.nc"
                )
                logger.info(
                    f"Wrote {config['fesom']['thisrun_forcing_dir'] + '/gfw_atmo.nc'} for use in the simulation!"
                )
                config["oasis3mct"].setdefault("coupling_input_fields", {})
                config["oasis3mct"]["coupling_input_fields"]["gfw_atmo"] = {
                    "freq": 86400,
                    "field_filepath": "gfw_atmo.nc",
                }
                config["fesom"].setdefault("forcing_files", {})
                config["fesom"].setdefault("forcing_sources", {})
                config["fesom"].setdefault("forcing_in_work", {})

                config["fesom"]["forcing_files"]["gfw_atmo"] = "gfw_atmo"
                config["fesom"]["forcing_sources"]["gfw_atmo"] = (
                    config["fesom"]["thisrun_forcing_dir"] + "/gfw_atmo.nc"
                )
                config["fesom"]["forcing_in_work"]["gfw_atmo"] = "gfw_atmo.nc"
                return config
            elif hosing_type == "from_file":
                logger.error("Sorry, hosing from a file has no yet been implemented!")
                sys.exit(1)
                # TODO(PG): Later...
                # hosing_file = config["fesom"]["hosing_file"]
            else:
                logger.error(
                    "Sorry, you must select between hosing_type: 'homogeneous' or hosing_type: 'from_file' in your fesom configuration!"
                )
                sys.exit(1)

    return config


gfw_creator/gfw_creator.py
# -*- coding: utf-8 -*-

"""
To use GFW Creator in a project::

    import gfw_creator

This gives you access to several functions, which are documented in detail in
the API section. The most important one is::

    >>> gfw_creator.create_homogenous_hosing
    <function create_homogeneous_hosing at 0x119820510>
    >>> # To use this, give lat/lon coordinates, and a hosing strength
    >>> hosing_ds = gfw_creator.create_homogeneous_hosing(45, 60, -30, -10, 0.15)
    >>> hosing_ds
    <xarray.Dataset>
    Dimensions:    (lat: 96, lon: 192)
    Coordinates:
      * lon        (lon) float64 0.0 1.875 3.75 5.625 ... 352.5 354.4 356.2 358.1
      * lat        (lat) float64 88.57 86.72 84.86 83.0 ... -84.86 -86.72 -88.57
    Data variables:
        cell_area  (lat, lon) float64 ...
        gfw_atmo   (lat, lon) float32 ...
    Attributes:
        CDI:          Climate Data Interface version 1.9.2 (http://mpimet.mpg.de/...
        Conventions:  CF-1.6
        history:      Wed Dec 23 14:28:30 2020: cdo -O -f nc -setclonlatbox,6.454...
        Comments:     Freshwater forcing file for use with the gfw_atmo switch in...
        Creators:     Dr. Paul Gierz (pgierz@awi.de)
        Institute:    Alfred Wegener Institute, Helmholtz Centre for Polar and Ma...
        CDO:          Climate Data Operators version 1.9.2 (http://mpimet.mpg.de/...

Since you get back `an xarray.Dataset object
<http://xarray.pydata.org/en/stable/generated/xarray.Dataset.html#xarray-dataset>`_,
you can easily continue working with this for further analysis or plotting.
"""

import cdo
import xarray as xr

import os
import pkgutil
import subprocess
import sys


def build_cdo_object():
    try:
        return cdo.Cdo()
    except Exception as e:
        print("Python exception:", e)
        print("Unable to use CDO, this won't work, sorry...")
        print("You may consider running >> module load cdo << and trying to run again")
        sys.exit(1)

def _get_template_file():
    """
    Creates a local copy of the gfw_atmo template file.

    Returns
    -------
    str :
        Path to the file that was generated
    """
    template_file_path = "./gfw_atmo_template.nc"
    template_file_data = pkgutil.get_data("gfw_creator", "data/gfw_atmo_template.nc")
    with open(template_file_path, "wb") as template_file:
        template_file.write(template_file_data)
    return os.path.abspath(template_file_path)


def _load_template_file():
    """
    Loads the template file to be modified.

    Returns
    -------
    xr.Dataset :
        The template dataset, has 0 everywhere with 366 day timesteps. This is
        on the standard ECHAM6 T63 grid.
    """
    df = xr.open_dataset(_get_template_file())
    df.coords["lon"] = (df.coords["lon"] + 180) % 360 - 180
    df = df.sortby(df.lon)
    return df


def _lat_lon_area(lat_0, lat_1, lon_0, lon_1):
    """
    Figures out the area of a lat/lon box for distributing a flux across a
    total area
    """
    CDO = build_cdo_object()
    ds = _get_template_file()
    selected_ds = CDO.fldsum(
        input=f"-sellonlatbox,{lon_0},{lon_1},{lat_0},{lat_1} -selname,cell_area {ds}",
        returnXDataset=True,
    )
    return selected_ds.cell_area.data.squeeze()


def create_homogeneous_hosing(lat_0, lat_1, lon_0, lon_1, hosing_strength):
    """
    Creates a homogeneous hosing field.

    Generates a uniform hosing field bounded by the box (``lat_0``, ``lon_0``)
    to (``lat_1``, ``lon_1``). The parameter ``hosing_strength`` is given in
    Sv, yet the resulting field has units m/s.

    Parameters
    ----------
    lat_0 : float
    lat_1 : float
    lon_0 : float
    lon_1 : float
    hosing_strength : float

    Returns
    -------
    xarray.Dataset
    """
    CDO = build_cdo_object()
    ds = _get_template_file()
    # Notes:
    # ------
    # `hosing_strength` is given in Sv. Convert this to m3/s:
    #
    # 1 Sv = 10**6 m3/s
    hosing_strength *= 1.0e6
    # The gfw_atmo file must be given in m/s, so we divide by the area for
    # conversion m3/s --> m/s :
    hosing_per_area = hosing_strength / _lat_lon_area(lat_0, lat_1, lon_0, lon_1)
    # Now that the hosing strength has been determined, set it in the NetCDF
    # file:
    gfw_atmo_file = CDO.setclonlatbox(
        f"{hosing_per_area},{lon_0},{lon_1},{lat_0},{lat_1}",
        input=ds,
        returnXDataset=True,
    )
    return gfw_atmo_file
